File: DataStructures/Trees/BinarySearchTree/BST.py
class BST:

    class Node:
        def __init__(self, key, val):
            self.key = key
            self.val = val
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None
    
    def put(self, key, val):
        self.root = self.__put_helper(self.root, key, val)

    def __put_helper(self, node, key, val):
        if node is None:
            return BST.Node(key, val)
        if key < node.key:
            node.left = self.__put_helper(node.left, key, val)
        elif key > node.key:
            node.right = self.__put_helper(node.right, key, val)
        else:  # key already exists, so we replace the value
            node.val = val
        return node
    
    def get(self, key):
        node = self.root
        while node is not None:
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                return node.val
        return None

    def get_max(self):
        return self.__get_max_helper(self.root)
    
    def __get_max_helper(self, node):
        if node.right is None:
            return node
        return self.__get_max_helper(node.right)

    def get_min(self):
        return self.__get_min_helper(self.root)
    
    def __get_min_helper(self, node):
        if node is None:
            return None
        elif node.left is None:
            return node
        else:
            return self.__get_min_helper(node.left)

    def delete_min(self):
        self.root = self.__delete_min_helper(self.root)

    def __delete_min_helper(self, node):
        if node.left is None:
            return node.right
        node.left = self.__delete_min_helper(node.left)
        return node

    def delete(self, key):
        self.root = self.__delete_helper(self.root, key)

    def __delete_helper(self, node, key):
        if node is None:
            return None
        if key < node.key:
            node.left = self.__delete_helper(node.left, key)
        elif key > node.key:
            node.right = self.__delete_helper(node.right, key)
        else:
            if node.right is None:  # Case where node has only a left child
                return node.left
            if node.left is None:  # Case where node has only a right child
                return node.right
            # Case where node has two children
            temp = node  # save reference to node
            node = self.__get_min_helper(temp.right)  # replace ref of node with successor
            node.right = self.__delete_min_helper(temp.right)  # delete successor from original place in tree
            node.left = temp.left  # put deleted nodes children onto successor
        return node
    
    def get_keys_in_order(self):
        keys = []
        self.__get_keys_in_order_helper(self.root, keys)
        return keys

    def __get_keys_in_order_helper(self, node, keys):
        if node is None:
            return
        if node.left is not None:
            self.__get_keys_in_order_helper(node.left, keys)
        keys.append(node.key)
        self.__get_keys_in_order_helper(node.right, keys)

    def get_vals_in_order(self):
        vals = []
        self.__get_vals_in_order_helper(self.root, vals)
        return vals

    def __get_vals_in_order_helper(self, node, vals):
        if node is None:
            return
        if node.left is not None:
            self.__get_vals_in_order_helper(node.left, vals)
        vals.append(node.val)
        self.__get_vals_in_order_helper(node.right, vals)
    

def main():

    test = {
        "56": "Bob",
        "100": "Joe",
        "78": "Fred",
        "1": "Al",
        "35": "Greg",
        "90": "Yori",
        "16": "Nola",
        "85": "Dia",
        "2": "Francis",
        "37": "Blob",
        "22": "Zebo"
    }

    bst = BST()

    for key, val in test.items():
        bst.put(int(key), val)

    bst.delete(16)
    bst.delete(2)

    print(bst.get_vals_in_order())
    print(bst.get_keys_in_order())
    print(bst.get(22))
    print(bst.get(100))

main()

File: DataStructures/Trees/BinarySearchTree/BST.js
class Node {
    constructor(key, val) {
        this.key = key
        this.val = val
        this.left = null
        this.right = null
    }
}

class BST {
    constructor() {
        this.root = null
        this.size = 0
    }

    put(key, val) {
        this.root = this.putHelper(this.root, key, val)
    }

    putHelper(node, key, val) {
        if (node === null) {  // we have found an empty slot to put our node
            this.size++
            return new Node(key, val)
        } 
        else if (key < node.key) node.left = this.putHelper(node.left, key, val)
        else if (key > node.key) node.right = this.putHelper(node.right, key, val)
        else node.val = val  // key === node.key
        return node
    }

    get(key) {
        let node = this.root
        while (node !== null) {
            if (key < node.key) node = node.left
            else if (key > node.key) node = node.right
            else return node.val
        }
        return null  // if no match found
    }

    delete(key) {
        this.root = this.deleteHelper(this.root, key)
    }

    deleteHelper(node, key) {
        if (node === null) return null
        else if (key < node.key) node.left = this.deleteHelper(node.left, key)
        else if (key > node.key) node.right = this.deleteHelper(node.right, key)
        else { // found node to delete
            this.size--
            if (node.left === null) return node.right // case where node has no left child
            if (node.right === null) return node.left // case where node has no right child
            // case where node has two children
            let temp = node
            node = this.getMinOfSubtree(temp.right) // replace node with successor
            node.right = this.deleteMinOfSubtree(temp.right)
            node.left = temp.left
        }
        return node
    }

    getMinOfSubtree(node) {
        while (node.left !== null) node = node.left
        return node
    }

    deleteMinOfSubtree(node) {
        if (node.left === null) return node.right
        node.left = this.deleteMinOfSubtree(node.left)
        return node
    }

    getSize() {
        return this.size
    }

    getSortedKeys() {
        const arr = []
        this.getSortedKeysHelper(this.root, arr)
        return arr
    }

    getSortedKeysHelper(node, arr) {
        if (node === null) return
        if (node.left !== null) this.getSortedKeysHelper(node.left, arr)
        arr.push(node.key)
        this.getSortedKeysHelper(node.right, arr)
    }
}


function test() {

    const data = {
        "56": "Bob",
        "100": "Joe",
        "78": "Fred",
        "1": "Al",
        "35": "Greg",
        "90": "Yori",
        "16": "Nola",
        "85": "Dia",
        "2": "Francis",
        "37": "Blob",
        "22": "Zebo"
    }

    const bst = new BST()
    
    for (let key in data) {
        bst.put(Number(key), data.key)
    }

    console.log(bst.getSize())

    bst.delete(90)
    bst.delete(1)

    console.log(bst.getSize())

    console.log(bst.getSortedKeys())
    console.log(bst.getSortedValues())
}


test()
